File: others/stripe/Order-Shipping-Cost/order_shipping_cost.py
from typing import Any, Dict, List, Optional


def _units_in_tier(quantity: int, min_quantity: int, max_quantity: Optional[int]) -> int:
    # Tier lower bound:
    # In this problem, minQuantity=0 means "start from unit 1".
    start = 1 if min_quantity <= 0 else min_quantity

    # Tier upper bound:
    # maxQuantity=None means this tier has no explicit upper bound,
    # so we cap it by current order quantity.
    end = quantity if max_quantity is None else max_quantity

    # Overlap length between [1, quantity] and [start, end].
    overlap_end = min(quantity, end)
    return max(0, overlap_end - start + 1)


def _calculate_product_cost(quantity: int, product_rule: Dict[str, Any]) -> int:
    if quantity == 0:
        return 0

    # 第一问: fixed per-item product cost.
    if "cost" in product_rule and "costs" not in product_rule:
        return quantity * product_rule["cost"]

    tiers: List[Dict[str, Any]] = product_rule.get("costs", [])

    total = 0

    for tier in sorted(tiers, key=lambda x: x.get("minQuantity", 0)):
        units = _units_in_tier(
            quantity,
            tier.get("minQuantity", 0),
            tier.get("maxQuantity"),
        )
        if units <= 0:
            continue

        tier_type = tier.get("type", "incremental")
        if tier_type == "fixed":
            total += tier["cost"]
        else:
            total += units * tier["cost"]

    return total


def calculate_shipping_cost(order: Dict[str, Any], shipping_costs: Dict[str, Any]) -> int:
    country = order["country"]
    product_to_rule = {rule["product"]: rule for rule in shipping_costs[country]}

    total = 0
    for item in order["items"]:
        product = item["product"]
        quantity = item["quantity"]
        total += _calculate_product_cost(quantity, product_to_rule[product])

    return total


if __name__ == "__main__":
    order_us = {
        "country": "US",
        "items": [
            {"product": "mouse", "quantity": 20},
            {"product": "laptop", "quantity": 5},
        ],
    }
    order_ca = {
        "country": "CA",
        "items": [
            {"product": "mouse", "quantity": 20},
            {"product": "laptop", "quantity": 5},
        ],
    }

    # 第一问: 固定单价
    shipping_cost_v1 = {
        "US": [
            {"product": "mouse", "cost": 550},
            {"product": "laptop", "cost": 1000},
        ],
        "CA": [
            {"product": "mouse", "cost": 750},
            {"product": "laptop", "cost": 1100},
        ],
    }
    assert calculate_shipping_cost(order_us, shipping_cost_v1) == 16000
    assert calculate_shipping_cost(order_ca, shipping_cost_v1) == 20500

    # 第二问: 按区间 incremental 逐段累计
    shipping_cost_v2 = {
        "US": [
            {
                "product": "mouse",
                "costs": [{"minQuantity": 0, "maxQuantity": None, "cost": 550}],
            },
            {
                "product": "laptop",
                "costs": [
                    {"minQuantity": 0, "maxQuantity": 2, "cost": 1000},
                    {"minQuantity": 3, "maxQuantity": None, "cost": 900},
                ],
            },
        ],
        "CA": [
            {
                "product": "mouse",
                "costs": [{"minQuantity": 0, "maxQuantity": None, "cost": 750}],
            },
            {
                "product": "laptop",
                "costs": [
                    {"minQuantity": 0, "maxQuantity": 2, "cost": 1100},
                    {"minQuantity": 3, "maxQuantity": None, "cost": 1000},
                ],
            },
        ],
    }
    assert calculate_shipping_cost(order_us, shipping_cost_v2) == 15700
    assert calculate_shipping_cost(order_ca, shipping_cost_v2) == 20200

    # 第三问: 区间内支持 fixed + incremental 混合
    shipping_cost_v3 = {
        "US": [
            {
                "product": "mouse",
                "costs": [
                    {
                        "type": "incremental",
                        "minQuantity": 0,
                        "maxQuantity": None,
                        "cost": 550,
                    }
                ],
            },
            {
                "product": "laptop",
                "costs": [
                    {"type": "fixed", "minQuantity": 0, "maxQuantity": 2, "cost": 1000},
                    {
                        "type": "incremental",
                        "minQuantity": 3,
                        "maxQuantity": None,
                        "cost": 900,
                    },
                ],
            },
        ],
        "CA": [
            {
                "product": "mouse",
                "costs": [
                    {
                        "type": "incremental",
                        "minQuantity": 0,
                        "maxQuantity": None,
                        "cost": 750,
                    }
                ],
            },
            {
                "product": "laptop",
                "costs": [
                    {"type": "fixed", "minQuantity": 0, "maxQuantity": 2, "cost": 1100},
                    {
                        "type": "incremental",
                        "minQuantity": 3,
                        "maxQuantity": None,
                        "cost": 1000,
                    },
                ],
            },
        ],
    }
    assert calculate_shipping_cost(order_us, shipping_cost_v3) == 14700
    assert calculate_shipping_cost(order_ca, shipping_cost_v3) == 19100

    print("All assertions passed.")

File: others/stripe/Order-Shipping-Cost/OrderShippingCost.go
package stripe

import (
	"fmt"
	"sort"
)

type Order struct {
	Country string      `json:"country"`
	Items   []OrderItem `json:"items"`
}

type OrderItem struct {
	Product  string `json:"product"`
	Quantity int    `json:"quantity"`
}

type ShippingCosts map[string][]ProductShippingRule

type ProductShippingRule struct {
	Product string         `json:"product"`
	Cost    *int           `json:"cost,omitempty"`
	Costs   []ShippingTier `json:"costs,omitempty"`
}

type ShippingTier struct {
	Type        string `json:"type,omitempty"`
	MinQuantity int    `json:"minQuantity"`
	MaxQuantity *int   `json:"maxQuantity"`
	Cost        int    `json:"cost"`
}

func CalculateShippingCost(order Order, shippingCosts ShippingCosts) (int, error) {
	countryRules, ok := shippingCosts[order.Country]
	if !ok {
		return 0, fmt.Errorf("unsupported country: %s", order.Country)
	}

	productToRule := make(map[string]ProductShippingRule, len(countryRules))
	for _, rule := range countryRules {
		productToRule[rule.Product] = rule
	}

	total := 0
	for _, item := range order.Items {
		rule, found := productToRule[item.Product]
		if !found {
			return 0, fmt.Errorf("missing shipping rule for product=%s country=%s", item.Product, order.Country)
		}
		productCost, err := calculateProductCost(item.Quantity, rule)
		if err != nil {
			return 0, err
		}
		total += productCost
	}

	return total, nil
}

func calculateProductCost(quantity int, rule ProductShippingRule) (int, error) {
	if quantity < 0 {
		return 0, fmt.Errorf("quantity must be non-negative")
	}
	if quantity == 0 {
		return 0, nil
	}

	// Variant 1: fixed per-item price.
	if rule.Cost != nil && len(rule.Costs) == 0 {
		return quantity * (*rule.Cost), nil
	}

	if len(rule.Costs) == 0 {
		return 0, fmt.Errorf("missing shipping cost tiers for product=%s", rule.Product)
	}

	tiers := append([]ShippingTier(nil), rule.Costs...)
	sort.Slice(tiers, func(i, j int) bool {
		return tiers[i].MinQuantity < tiers[j].MinQuantity
	})

	total := 0
	matched := false
	for _, tier := range tiers {
		units := unitsInTier(quantity, tier.MinQuantity, tier.MaxQuantity)
		if units <= 0 {
			continue
		}
		matched = true

		tierType := tier.Type
		if tierType == "" {
			tierType = "incremental"
		}

		switch tierType {
		case "fixed":
			total += tier.Cost
		case "incremental":
			total += units * tier.Cost
		default:
			return 0, fmt.Errorf("unknown tier type: %s", tierType)
		}
	}

	if !matched {
		return 0, fmt.Errorf("no pricing tier matched quantity=%d product=%s", quantity, rule.Product)
	}
	return total, nil
}

func unitsInTier(quantity, minQuantity int, maxQuantity *int) int {
	if quantity <= 0 {
		return 0
	}

	start := tierStart(minQuantity)
	if quantity < start {
		return 0
	}

	end := quantity
	if maxQuantity != nil {
		end = *maxQuantity
	}
	if end < start {
		return 0
	}

	if end > quantity {
		end = quantity
	}
	return end - start + 1
}

func tierStart(minQuantity int) int {
	if minQuantity <= 0 {
		return 1
	}
	return minQuantity
}
